pydicom/color.py

```diff
@@ -0,0 +1,149 @@
+
+            Raises
+            ------
+            ValueError
+                When ICC Profile cannot be read.
+            RuntimeError
+                When color management is not possible due to missing
+                dependencies.
+
+            """
+            try:
+                self._icc_transform = self._build_icc_transform(icc_profile)
+            except OSError:
+                raise ValueError('Could not read ICC Profile.')
+
+        def transform_frame(self, array: np.ndarray) -> np.ndarray:
+            """Transforms a frame by applying the ICC profile.
+
+            Parameters
+            ----------
+            array: numpy.ndarray
+                Pixel data of a color image frame in form of an array with
+                dimensions (Rows x Columns x SamplesPerPixel)
+
+            Returns
+            -------
+            numpy.ndarray
+                Color corrected pixel data of a image frame in form of an array
+                with dimensions (Rows x Columns x SamplesPerPixel)
+
+            Raises
+            ------
+            ValueError
+                When `array` does not have 3 dimensions and thus does not represent
+                a color image frame.
+
+            """
+            if array.ndim != 3:
+                raise ValueError(
+                    'Array has incorrect dimensions for a color image frame.'
+                )
+            image = Image.fromarray(array)
+            applyTransform(image, self._icc_transform, inPlace=True)
+            return np.asarray(image)
+
+        @staticmethod
+        def _build_icc_transform(icc_profile: bytes) -> ImageCmsTransform:
+            """Builds an ICC Transformation object.
+
+            Parameters
+            ----------
+            icc_profile: bytes
+                ICC Profile
+
+            Returns
+            -------
+            PIL.ImageCms.ImageCmsTransform
+                ICC Transformation object
+
+            """
+            profile: bytes
+            try:
+                profile = ImageCmsProfile(BytesIO(icc_profile))
+            except OSError:
+                raise ValueError('Cannot read ICC Profile in image metadata.')
+            name = getProfileName(profile).strip()
+            description = getProfileDescription(profile).strip()
+            logger.debug(f'found ICC Profile "{name}": "{description}"')
+
+            logger.debug('build ICC Transform')
+            intent = ImageCms.INTENT_RELATIVE_COLORIMETRIC
+            if not isIntentSupported(
+                profile,
+                intent=intent,
+                direction=ImageCms.DIRECTION_INPUT
+            ):
+                raise ValueError(
+                    'ICC Profile does not support desired '
+                    'color transformation intent.'
+                )
+            return ImageCms.buildTransform(
+                inputProfile=profile,
+                outputProfile=ImageCms.createProfile('sRGB'),
```

Comment:
Maybe add a comment that this is the only supported color space, but others may be used

Reply:
sRGB is a very limited color space and its use as an intermediary may result in out of gamut colors being inappropriately handled - how is your color management implementation going to take advantage of a calibrated monitor (especially a "wide gamut" monitor), or in any case, with a gamut wider than sRGB supports (not to mention the quantization issues caused by using 8 bit per channel intermediary spaces for transformations that are not linear)?

I appreciate that anything beyond an sRGB working space in HTML5 Canvas may be a bit beyond the state of the art at the moment, but if this is not handled properly, images in browsers may display poorly compared to thick clients; see also [https://lists.w3.org/Archives/Public/public-colorweb/2021Dec/att-0002/2021-Conversion_Overview.pdf](https://lists.w3.org/Archives/Public/public-colorweb/2021Dec/att-0002/2021-Conversion_Overview.pdf).

You might want to consult with some color experts from ICC, W3C and FDA.

Reply:
Thanks for your feedback @dclunie. For the Python library, my focus has been on image analysis (for which sRGB may be ok?) rather than display use cases. However, I agree that we should implement this in a way that would work for a variety of use cases.

@scaramallion we could just add an optional parameter `output_profile` to allow the caller to provide the output ICC profile as well.

Reply:
Since the caller can already provide the transform object, we may indeed just need to document the default behaviour.

Reply:
Sorry, I wasn't very clear. Of the three possible output colour spaces currently specified by the Standard (sRGB, Adobe, ROMM), it appears [Pillow only supports sRGB](https://pillow.readthedocs.io/en/stable/reference/ImageCms.html#PIL.ImageCms.createProfile).